### ISSUE TITLE
Coordinate ptem with variable `opsz`

### ISSUE THREAD

**Issue by ebraminio** [harfbuzz/harfbuzz#1914]
Great that we now have #1484 as it will ease this, one thing I think worth to do eventually is to coordinate _set_ptem with `opsz` variation axis (and maybe deprecate _set_ptem eventually), I am not sure how that should be done, maybe simulating an axis variation when `trak` table exists? Setting the axis when _set_ptem is called? etc. Let me know what do you think

**Comment by behdad** [comment]
I agree we want some of those things.  Not sure about specifics.  Let this sit here for a while. :)  I thought I opened an issue for it before.

**Comment by drott** [comment]
I am concerned about the case that in Blink it would be possible to override `opsz` using font-variation-settings and disconnect it from ptem-size - in that case I prefer it not to be automatic behind the scenes.

**Comment by behdad** [comment]
So I think `trak` and `opsz` should be controlled the same way.  Same way that we control TT `kern` table using `kern` OT feature.

How to control it, what takes priority over what, to be determined, and is entangled in other things, like named-instances and desire that their variations stick. Ie. https://github.com/harfbuzz/harfbuzz/issues/1883

So, we're moving to a place that a font's variations don't start to default coordinates, but they come from the named-instance, then overriden by `set_pptm` and only then overriden by explicitly-specified varitations.

**Comment by behdad** [comment]
So Blink can always override.

**Comment by matthiasclasen** [comment]
This would be great to have. I'm looking at what to do for opsz in pango, and having harfbuzz just do the right thing would be by far the easiest.

**Comment by behdad** [comment]
> This would be great to have. I'm looking at what to do for opsz in pango, and having harfbuzz just do the right thing would be by far the easiest.

I agree. As a first step let's make pango call `hb_font_set_ptem` if it doesn't.

**Comment by behdad** [comment]
Ref https://gitlab.gnome.org/GNOME/pango/-/issues/621

**Comment by behdad** [comment]
Thinking more about this I think I want to deprecate this API and make it simply return the currently-set value of the `opsz` axis. Adding two new API:
```
  void hb_font_set_variation(font, hb_tag_t, float value);
  float hb_font_get_variation(font, hb_tag_t);
```
these set/get one variation without disturbing others.

The reasoning is that if we were to keep the ptem API, then we'd also need to add API for weight and width, and other standard axes when they are added. But by precedent we prefer enum-based APIs like our hb-style, hb-ot-metrics, etc.  That API style also obviates the interaction questions raised.

**Comment by behdad** [comment]
> Thinking more about this I think I want to deprecate this API and make it simply return the currently-set value of the `opsz` axis. Adding two new API:
> 
> ```
>   void hb_font_set_variation(font, hb_tag_t, float value);
>   float hb_font_get_variation(font, hb_tag_t);
> ```
> 
> these set/get one variation without disturbing others.
> 
> The reasoning is that if we were to keep the ptem API, then we'd also need to add API for weight and width, and other standard axes when they are added. But by precedent we prefer enum-based APIs like our hb-style, hb-ot-metrics, etc. That API style also obviates the interaction questions raised.

@jfkthame does this make sense to you?

**Comment by behdad** [comment]
Seems like this is the root of a few bugs:

https://github.com/harfbuzz/harfbuzz/issues/1914
https://gitlab.gnome.org/GNOME/pango/-/issues/695

**Comment by khaledhosny** [comment]
We already have `hb_font_set_variation()`/`hb_font_get_variation()`. I don’t think we want to coordinate `ptem` with `opsz` (it is too late for such a breaking change). Please re-open if you think there is anything left that needs to be done here.
